### ISSUE TITLE
Add convert(x, to = "kerasR") functionality

### ISSUE THREAD

**Issue by kbenoit** [quanteda/quanteda#1145]
Following the discussion in #1138, we are thinking of extending `convert()` so that we could go from a dfm using:

```r
convert(anydfm, to = "kerasR")
```

Problem is we haven't used keras or **kerasR**.

@statsmaths can you tell us what format of object would be the best to provide from **quanteda**? 

Would extending `convert()` be most appropriate? We can also provide conversion from a tokens object, but it seems more natural to convert a dfm.  Would/should it possibly include additional document variables? 

We're ready to implement it but would appreciate the guidance.

**Comment by statsmaths** [comment]
I think that is a great idea! Plus, the exact same output could be used with the **keras** package (R studio's port to make keras more "R like") and my **kerasR** (which follows the Python interface more closely).

Ultimately, keras wants a matrix with one row per document and one column for the maximum length (in tokens) of a given document. As an intermediate step, you'll need a list with a single element per document. Each element contains a vector of (ordered) word indices. Both packages provide `pad_sequences`, which allows for going from the intermediate step to the final version.

To give you a working example, here  is the first step using **tokenizers**:

```{r}
docs <- c("I like R.", "I really like quanteda.", "I love fasttext")
tokens <- tokenizers::tokenize_words(docs)
vocab <- unique(unlist(tokens))
intermediate_results <- sapply(tokens, function(v) match(v, vocab))
intermediate_results
```
```
[[1]]
[1] 1 2 3

[[2]]
[1] 1 4 2 5

[[3]]
[1] 1 6 7
```

Which vocabulary element gets mapped to an index is irrelevant to keras, though I assume users would want to have access to the object `vocab` (it is also needed if someone is applying the same model to a new dataset). Once someone has this list, they can quickly convert to the format needed in keras:

```{r}
keras_input <- kerasR::pad_sequences(intermediate_results)
keras_input
```
```
     [,1] [,2] [,3] [,4]
[1,]    0    1    2    3
[2,]    1    4    2    5
[3,]    0    1    6    7
```

I would recommend returning just `intermediate_results` in **quanteda**. There are some important  options in `kerasR::pad_sequences`/`keras::pad_sequences` that users will likely want to adjust.

I hope that helps describe the output format. If you'd like some help with building the functionality, just let me know.

**Comment by kbenoit** [comment]
OK, so it sounds like we have implemented this in #1141. This converts a **quanteda** `tokens` object into what you have called here `intermediate_results`. We cannot implement the options in `kerasR::pad_sequences()` from a dfm object, since that has lost the word order.

If the `intermediate_results` object is not useful directly for anything, however, we could simply implement a `pad_sequences.tokens()` method and then no conversion would be required.

Thoughts?

**Comment by FarzanT** [comment]
Sorry if I misunderstand, but using the function `as.keras.tokens()` in #1141 seems to not maintain the word order:
```R
> if (!require(quanteda)) {
+     install.packages("quanteda")
+     library(quanteda)
+ }
> text1 <- c(old1 = "I love the quanteda package")
> text2 <- c(old2 = "I love the R language, the R language is great")
> text3 <- c(old3 = "R is so great")
> as.keras.tokens <- function(x) {
+         unname(unclass(x))
+ }
> trainTexts = c(text1, text2, text3)
> myTokens <- tokens(trainTexts)
> kerasTokens <- as.keras.tokens(myTokens)
> kerasTokens[1:3]
[[1]]
[1] 1 2 3 4 5

[[2]]
 [1]  1  2  3  6  7  8  3  6  7  9 10

[[3]]
[1]  6  9 11 10
```

**Comment by FarzanT** [comment]
May I add that the vocabulary/mapping associated with the intermediate object would be very useful in indexing/tokenizing a new dataset (as mentioned above). I hope this can be achieved if intermediate results are not kept.

**Comment by koheiw** [comment]
We have `types.tokens()` to extract vocabulary, so we will just pass it to `tokens()` to index words in new texts in the same way.

**Comment by FarzanT** [comment]
Thank you!
Just making sure that I'm doing this correctly:
```R
> text1 <- c(old1 = "I love the quanteda package")
> text2 <- c(old2 = "I love the R language, the R language is great the the")
> text3 <- c(old3 = "R is so great")
> toks1 <- tokens(c(text1, text2))
> (toks2 <- tokens(text3))
tokens from 1 document.
old3 :
[1] "R"     "is"    "so"    "great"

> indices <- unname(unclass(toks1))
> pat <- types(toks1)
> (rep <- as.character(1:length(pat)))
 [1] "1"  "2"  "3"  "4"  "5"  "6"  "7"  "8"  "9"  "10"
> # Remove unseen words
> (toks2 <- tokens_select(x = toks2, pattern = pat))
tokens from 1 document.
old3 :
[1] "R"     "is"    "great"

> # Replace tokens with their indices
> tokens_replace(x = toks2, pattern = pat, replacement = rep)
tokens from 1 document.
old3 :
[1] "6"  "9"  "10"
```
I wish there was a single function/argument to do all of this, would make it more foolproof.

**Comment by statsmaths** [comment]
This generally looks like the right approach. Personally, I am okay with the package just returning the list of indices. As mentioned above, there are usual options to the padding sequences bit that users may want to play with (such as how long to pad/truncate by). Also, many people may want to filter to only the top N terms. Using the inverse word frequency as an index, as you seem to here, makes this easy.

As mentioned above and in #1141, having a way to pass a pre-made vocabulary into the method would be extremely helpful. Words not in the vocabulary can either by coded as `0` or ignored.

**Comment by koheiw** [comment]
This is only for experimental purpose as it is untested and inefficient, but we could do 
```r
text1 <- c(old1 = "I love the quanteda package")
text2 <- c(old2 = "I love the R language, the R language is great the the")

toks <- tokens(text1)
unclass(toks)

# $old1
# [1] 1 2 3 4 5
# 
# attr(,"types")
# [1] "I"        "love"     "the"      "quanteda" "package" 

toks2 <- quanteda:::tokens_serialize(as.list(tokens(text2)), types(toks))
unclass(toks2)

# $old2
# [1]  1  2  3  6  7  8  3  6  7  9 10  3  3

# attr(,"types")
# [1] "I"        "love"     "the"      "quanteda" "package"  "R"        "language" ","        "is"       "great" 
```

**Comment by kbenoit** [comment]
See the discussion in PR #1141, which we will reopen when we return to this issue.

**Comment by konradsemsch** [comment]
Hello @kbenoit,

on first look and after checking a couple of scenarios I believe both of these should work. Especially now with ```tokens_to_kerassequences``` it would help a lot to pre-processed the tokens object before transforming it into sequences.

There's, however, one scenario that I see could be missing here but I'm not sure about the feasibility:
with ```dfm``` the user has the possibility to trim the tokens based on the frequencies defined in different ways. Such option is not available for ```tokens``` but it could be found very useful for some applications.

Is there also a possibility fo going back from a trimmed dfm to tokens so that they then could be converted into keras sequences?

**Comment by kbenoit** [comment]
No, because the dfm would erase the sequence of words. (Although removing them based on frequency destroys the original sequence too!) However if you simply want them removed from the keras-formatted tokens stream, this is possible to add in the function.

What would be the options to add to `tokens_to_kerassequences()` (which we will almost certainly rename when we add it to the package, eventually)? Please either describe them or point us to the equivalent arguments in a **keras** function (e.g. `keras::text_tokenizer()`).

**Comment by amatsuo** [comment]
From what I read, the data preparations in the example above by @jlacko can be replicated by this code in `quanteda` (and the following code in the example works with minimal modification).  

```r
tweets <- read.csv(tf_tweets, stringsAsFactors = F) 

corp <- corpus(tweets)
tok <- corp %>% tokens()
mt <- dfm(tok)
tok_to_keep <- mt %>% dfm_trim(min_docfreq = 5) %>% colnames()
tok <- tokens_select(tok, tok_to_keep, selection = "keep", 
                     valuetype = "fixed", padding = TRUE)

vocabulary <- data.frame(token = types(tok), tok_id = types(tok))

j <- tok %>% unclass %>%
  lapply(function(x) seq(length(x))) %>% unlist()
i <- tok %>% unclass %>% sapply(length) %>% rep(seq_along(.), times = .)
word_matrix <- Matrix::sparseMatrix(i = i, j = j, x = unlist(unclass(tok))) %>% as.matrix()

keras_input <- bind_cols(docvars(corp), text = texts(corp), as_data_frame(word_matrix))
```

So the implementation of `tokens_to_kerassequences ` may not be hard, but the question would be what kind of arguments there should be. For example, if the cutoff by docfreq is condcuted (as in the code above),  should it be done in the function, or should we expect users to create `dfm` and list the words to keep (and the users will supply the list of words as an vocaburary argument of the function).

**Comment by amatsuo** [comment]
@jlacko Thanks for the feedback. I now understand that `vocabulary` option is crucial for adding texts to the text set. Since that's the case, we will definitely have an option for providing the list of words.

**Comment by kbenoit** [comment]
Hi all, picking up this issue since we are working on a prototype in a new testing package [**quanteda.classifiers**](https://github.com/quanteda/quanteda.classifiers) that will make fitting and predicting work in the same way that `textmodel_nb()` currently works (meaning: easy! and as close to `lm()` as possible where `x` is a dfm and `y` is a vector of training labels that can include `NA` to be ignored by the model but kept in the model matrix for later prediction.) Why am I obsessed by this? Because users should be able to fit a ML predictive model for text as easily as they can run OLS (ok maybe with a few more options, and yeah, we are still thinking about a really simple way to make cross-validation work), without having to study complicated blog examples (although @jlacko I thought yours was really clear!) or work out how to use an API designed by technical people 🤓 for use by technical people.

It looks to me that in **keras**, `fit.keras.engine.training.Model()` accepts for `x` a "Vector, matrix, or array of training data". This looks like a dense object to me. Anyone have any ideas about how to get it to accept a sparse matrix object? Otherwise I can see that some extreme trimming is going to be needed.

**Comment by claravdw** [comment]
Was anything like this ever implemented? Thanks!

**Comment by kbenoit** [comment]
No, not yet. But relates to #2436, which might have now surpassed this issue in terms of technology and demand.
